File: backend/app/providers/mangakatana.py
"""
MangaKatana provider — HTML scraping via BeautifulSoup.
Fingerprints are checked before every scraping session to detect site layout changes.
"""
import re
import json
import asyncio
from bs4 import BeautifulSoup  # pyright: ignore[reportMissingImports]
from app.providers.base import (
    Provider, MangaResult, MangaDetail, ChapterResult,
    HealthReport, ProviderHealth, ScraperFingerprint,
)

SITE = "https://mangakatana.com"

# These selectors are checked on each validate() call.
_FINGERPRINTS = [
    ScraperFingerprint(name=".manga-list, .manga_list-sbs", critical=True), # search results grid
    ScraperFingerprint(name=".chapters, #chapters", critical=True),       # chapter list
]


class MangaKatanaProvider(Provider):
    id = "mangakatana"
    name = "MangaKatana"
    base_url = SITE
    fingerprints = _FINGERPRINTS

    async def validate(self) -> HealthReport:
        """
        Validate by hitting the search page and a known manga page.
        """
        failures: list[str] = []
        critical_failure = False

        try:
            client = await self._get_client()
            # Check search results page - use a generic search
            resp = await client.get(SITE, params={"search": "one", "search_by": "m_name"})
            resp.raise_for_status()
            soup = BeautifulSoup(resp.text, "html.parser")

            if not soup.select_one(".manga-list, .manga_list-sbs, .item"):
                failures.append("manga-list (search results)")
                # only critical if no fallback works
                if not soup.select(".item"):
                    critical_failure = True

            # Check a known stable manga page
            resp2 = await client.get(f"{SITE}/manga/one-piece.49")
            if resp2.status_code == 200:
                soup2 = BeautifulSoup(resp2.text, "html.parser")
                if not soup2.select_one(".chapters, #chapters, .chapter"):
                    failures.append("chapters (chapter list)")
                    critical_failure = True
        except Exception as exc:
            report = HealthReport(status=ProviderHealth.BROKEN, failures=["network_error"], message=str(exc))
            self._health = ProviderHealth.BROKEN
            self._health_report = report
            return report

        status = ProviderHealth.BROKEN if critical_failure else (ProviderHealth.DEGRADED if failures else ProviderHealth.OK)
        report = HealthReport(
            status=status,
            failures=failures,
            message=f"{len(failures)} selector(s) missing" if failures else "All checks passed",
        )
        self._health = status
        self._health_report = report
        return report

    async def search(self, query: str, page: int = 1) -> list[MangaResult]:
        client = await self._get_client()
        # Search is on the root URL or /manga
        # Based on user feedback: https://mangakatana.com/?search=onepiece&search_by=m_name
        url = f"{SITE}/page/{page}" if page > 1 else SITE
        resp = await client.get(url, params={
            "search": query,
            "search_by": "m_name",
        })
        
        # If the site redirects directly to a manga page (e.g. searching "one piece" might redirect to /manga/one-piece.49)
        if "/manga/" in str(resp.url) and resp.url != url:
            # We are on a detail page, extract info from here
            soup = BeautifulSoup(resp.text, "html.parser")
            slug = str(resp.url).rstrip("/").split("/")[-1]
            title_el = soup.select_one("h1.heading")
            img = soup.select_one(".cover img")
            return [MangaResult(
                id=slug,
                title=title_el.get_text(strip=True) if title_el else slug,
                cover_url=img.get("src") or img.get("data-src") if img else None,
                provider=self.id,
                url=str(resp.url),
            )]

        resp.raise_for_status()
        soup = BeautifulSoup(resp.text, "html.parser")

        results = []
        # Search results are usually in a div with .manga_list-sbs or just .item
        items = soup.select(".manga_list-sbs .item") or soup.select(".item")
        for item in items:
            a = item.select_one("h3 a") or item.select_one(".title a") or item.select_one("a")
            if not a or "/manga/" not in a["href"]:
                continue
            
            img = item.select_one("img")
            href = a["href"]
            slug = href.rstrip("/").split("/")[-1]
            
            if any(r.id == slug for r in results):
                continue

            results.append(MangaResult(
                id=slug,
                title=a.get_text(strip=True),
                cover_url=img.get("src") or img.get("data-src") if img else None,
                provider=self.id,
                url=href,
                status=self._parse_status(item),
            ))
        return results

    async def get_manga(self, manga_id: str) -> MangaDetail:
        client = await self._get_client()
        resp = await client.get(f"{SITE}/manga/{manga_id}")
        resp.raise_for_status()
        soup = BeautifulSoup(resp.text, "html.parser")

        title_el = soup.select_one("h1.heading")
        title = title_el.get_text(strip=True) if title_el else manga_id

        cover_el = soup.select_one(".cover img")
        cover = cover_el["src"] if cover_el else None

        desc_el = soup.select_one(".summary p") or soup.select_one(".summary")
        desc = desc_el.get_text(strip=True) if desc_el else None

        status = None
        for row in soup.select(".info-list .item"):
            label = row.select_one(".label")
            if label and "status" in label.get_text(strip=True).lower():
                val = row.select_one("span:not(.label)")
                if val:
                    status = val.get_text(strip=True).lower()

        genres = [a.get_text(strip=True) for a in soup.select(".genres a")]
        authors = [a.get_text(strip=True) for a in soup.select(".author a")]

        chapters = self._parse_chapters(soup, manga_id)

        return MangaDetail(
            id=manga_id,
            title=title,
            cover_url=cover,
            description=desc,
            status=status,
            genres=genres,
            authors=authors,
            provider=self.id,
            url=f"{SITE}/manga/{manga_id}",
            chapters=chapters,
        )

    async def get_pages(self, chapter_id: str) -> list[str]:
        # chapter_id format: "{manga_slug}/{chapter_slug}"
        client = await self._get_client()
        resp = await client.get(f"{SITE}/manga/{chapter_id}")
        resp.raise_for_status()
        soup = BeautifulSoup(resp.text, "html.parser")

        # MangaKatana stores image list in a JS variable: var thzq=[...] or similar
        # Multiple arrays might exist (for different servers). We pick the longest one.
        import ast
        scripts = soup.find_all("script")
        all_url_lists = []
        
        for script in scripts:
            if script.string:
                # Look for patterns like var \w+ = ['...', '...']
                matches = re.findall(r'var\s+\w+\s*=\s*(\[.*?\])\s*;', script.string, re.DOTALL)
                for content in matches:
                    try:
                        # Convert JS-style array with single quotes to Python list
                        urls = ast.literal_eval(content)
                        if isinstance(urls, list) and urls and isinstance(urls[0], str) and "http" in urls[0]:
                            all_url_lists.append(urls)
                    except (ValueError, SyntaxError):
                        pass

        if all_url_lists:
            # Return the longest list of URLs found
            return max(all_url_lists, key=len)

        # Fallback: look for img tags
        imgs = soup.select(".wrap_warpper img[data-src], .chapter-img img, #img_list img")
        if imgs:
            return [img.get("data-src") or img.get("src") for img in imgs if img.get("data-src") or img.get("src")]

        return []

    def _parse_status(self, item) -> str | None:
        badge = item.select_one(".status") or item.select_one(".chapter-count")
        if badge:
            text = badge.get_text(strip=True).lower()
            if "ongoing" in text:
                return "ongoing"
            if "completed" in text or "complete" in text:
                return "completed"
        return None

    def _parse_chapters(self, soup: BeautifulSoup, manga_id: str) -> list[ChapterResult]:
        chapters = []
        # MangaKatana uses a table inside .chapters
        for row in soup.select(".chapters tr"):
            a = row.select_one("a")
            if not a:
                continue
            href = a["href"]
            slug = href.rstrip("/").split("/")[-1]
            title = a.get_text(strip=True)

            # Extract chapter number from slug like "c001" or "chapter-5"
            num_match = re.search(r"[\d.]+", slug)
            num = float(num_match.group()) if num_match else 0.0

            date_el = row.select_one(".update_time") or row.select_one("time") or row.select_one("td:last-child")
            published = date_el.get_text(strip=True) if date_el else None

            chapters.append(ChapterResult(
                id=f"{manga_id}/{slug}",
                title=title,
                number=num,
                url=href,
                published_at=published,
            ))
        return chapters

File: backend/app/services/extensions/mangakatana.js
var _MK = 'https://mangakatana.com';

async function _fetchDoc(url) {
  var data = await apiFetch('/manga/proxy/html?url=' + encodeURIComponent(url));
  return new DOMParser().parseFromString(data.html, 'text/html');
}

async function _fetchHtml(url) {
  var data = await apiFetch('/manga/proxy/html?url=' + encodeURIComponent(url));
  return data.html;
}

var extension = {
  async search(query, page) {
    var pageStr = (page || 1) > 1 ? ('/page/' + page) : '';
    var doc = await _fetchDoc(_MK + pageStr + '?search=' + encodeURIComponent(query) + '&search_by=m_name');
    var results = [];
    var seen = {};
    doc.querySelectorAll('#book_list .item, .manga_list-sbs .item, .item').forEach(function(item) {
      var a = item.querySelector('h3.title a, .text h3 a, .title a, a');
      if (!a) return;
      var href = a.getAttribute('href') || '';
      if (!href.includes('/manga/')) return;
      var slug = href.replace(/\/$/, '').split('/').pop();
      if (!slug || seen[slug]) return;
      seen[slug] = true;
      var img = item.querySelector('.media .wrap_img img, img');
      results.push({
        id: slug,
        title: a.textContent.trim(),
        cover_url: img ? (img.getAttribute('src') || img.getAttribute('data-src')) : null,
        provider: 'mangakatana',
        url: href,
        status: null,
      });
    });
    return results;
  },

  async getMangaDetail(mangaId) {
    var doc = await _fetchDoc(_MK + '/manga/' + mangaId);
    var title = (doc.querySelector('h1.heading') || {textContent:mangaId}).textContent.trim();
    var coverEl = doc.querySelector('.cover img');
    var cover = coverEl ? (coverEl.getAttribute('src') || coverEl.getAttribute('data-src') || coverEl.getAttribute('data-lazy-src')) : null;
    var desc = null;
    var descEl = doc.querySelector('.summary p') || doc.querySelector('.summary');
    if (descEl) desc = descEl.textContent.trim();
    var genres = [];
    doc.querySelectorAll('.genres a').forEach(function(a) { genres.push(a.textContent.trim()); });
    var authors = [];
    doc.querySelectorAll('.author a').forEach(function(a) { authors.push(a.textContent.trim()); });

    var chapters = [];
    doc.querySelectorAll('.chapters tr').forEach(function(row) {
      var a = row.querySelector('a');
      if (!a) return;
      var href = a.getAttribute('href') || '';
      var slug = href.replace(/\/$/, '').split('/').pop();
      var numMatch = slug.match(/([\d.]+)/);
      var num = numMatch ? parseFloat(numMatch[1]) : 0;
      var dateEl = row.querySelector('.update_time, time, td:last-child');
      chapters.push({
        id: mangaId + '/' + slug,
        title: a.textContent.trim() || ('Chapter ' + num),
        number: num,
        published_at: dateEl ? dateEl.textContent.trim() : null,
      });
    });

    return {
      id: mangaId, title: title,
      cover_url: cover, description: desc,
      status: null, genres: genres, authors: authors,
      provider: 'mangakatana',
      url: _MK + '/manga/' + mangaId,
      chapters: chapters,
    };
  },

  async getPages(chapterId) {
    var html = await _fetchHtml(_MK + '/manga/' + chapterId);
    var best = [];
    var re = /var\s+\w+\s*=\s*(\['[^']*'(?:,'[^']*')*,?\])\s*;/g;
    var m;
    while ((m = re.exec(html)) !== null) {
      try {
        var str = m[1].replace(/'/g, '"').replace(/,\s*\]/, ']');
        if (!str.endsWith(']')) str += ']';
        var arr = JSON.parse(str);
        if (Array.isArray(arr) && arr.length > best.length && typeof arr[0] === 'string' && arr[0].startsWith('http')) {
          best = arr;
        }
      } catch(e) {}
    }
    if (best.length) return best;

    var doc = new DOMParser().parseFromString(html, 'text/html');
    var pages = [];
    doc.querySelectorAll('.wrap_warpper img[data-src], #img_list img').forEach(function(img) {
      var src = img.getAttribute('data-src') || img.getAttribute('src');
      if (src) pages.push(src);
    });
    return pages;
  },

  async getPopular(page) {
    var doc = await _fetchDoc(_MK + '/manga/page/' + (page || 1));
    var results = [];
    var seen = {};
    doc.querySelectorAll('#book_list .item, .manga_list-sbs .item, .item').forEach(function(item) {
      var a = item.querySelector('h3.title a, .text h3 a, .title a, a');
      if (!a || !a.getAttribute('href').includes('/manga/')) return;
      var slug = a.getAttribute('href').replace(/\/$/, '').split('/').pop();
      if (!slug || seen[slug]) return;
      seen[slug] = true;
      var img = item.querySelector('.media .wrap_img img, img');
      results.push({
        id: slug,
        title: a.textContent.trim(),
        cover_url: img ? (img.getAttribute('src') || img.getAttribute('data-src')) : null,
        provider: 'mangakatana',
        url: a.getAttribute('href'),
        status: null,
      });
    });
    return results;
  },

  async getLatest(page) {
    var doc = await _fetchDoc(_MK + '/manga/page/' + (page || 1) + '?filter=latest');
    var results = [];
    var seen = {};
    doc.querySelectorAll('#book_list .item, .manga_list-sbs .item, .item').forEach(function(item) {
      var a = item.querySelector('h3.title a, .text h3 a, .title a, a');
      if (!a || !a.getAttribute('href').includes('/manga/')) return;
      var slug = a.getAttribute('href').replace(/\/$/, '').split('/').pop();
      if (!slug || seen[slug]) return;
      seen[slug] = true;
      var img = item.querySelector('.media .wrap_img img, img');
      results.push({
        id: slug,
        title: a.textContent.trim(),
        cover_url: img ? (img.getAttribute('src') || img.getAttribute('data-src')) : null,
        provider: 'mangakatana',
        url: a.getAttribute('href'),
        status: null,
      });
    });
    return results;
  },
};
